### ISSUE TITLE
False TS2345 error for XOR type argument

### ISSUE THREAD

**Issue by danielzlotnik** [microsoft/TypeScript#31395]
**TypeScript Version:**  3.4.5

**Code**

```ts
type AxorB = {
  a: string; b?: never
} | {
  a?: never; b: string
};
    
function foo(bar: string) {}
    
function callFoo({a, b}: AxorB) {
  // TS2345: Argument of type 'string | undefined' is not assignable to parameter of type 'string'.
  foo(a || b);
}
```

**Expected behavior:**
No error - since either a or b must have a string value as implied by `AxorB` type.

**Actual behavior:**
Error

**Playground Link:** [link](https://www.typescriptlang.org/play/#src=%20%20%20%20type%20AxorB%20%3D%20%7B%0D%0A%20%20%20%20%20%20a%3A%20string%3B%20b%3F%3A%20never%0D%0A%20%20%20%20%7D%20%7C%20%7B%0D%0A%20%20%20%20%20%20a%3F%3A%20never%3B%20b%3A%20string%0D%0A%20%20%20%20%7D%3B%0D%0A%20%20%20%20%0D%0A%20%20%20%20function%20foo(bar%3A%20string)%20%7B%0D%0A%20%20%20%20%20%20%2F%2F%20whatever%0D%0A%20%20%20%20%7D%0D%0A%20%20%20%20%0D%0A%20%20%20%20function%20callFoo(%7Ba%2C%20b%7D%3A%20AxorB)%20%7B%0D%0A%20%20%20%20%20%20foo(a%20%7C%7C%20b)%3B%0D%0A%20%20%20%20%7D) - turn on `strictNullChecks` to see the error.

**Comment by nmain** [comment]
I don't think `tsc` tracks this sort of thing, but even if it did the example is still incorrect:

```ts
callFoo({ a: "" }) // at runtime, will attempt to call foo(undefined)
```

**Comment by danielzlotnik** [comment]
*Facepalm* that's indeed it - closing.
